[Callbacks] DOC Add callbacks developer documentation (#33423)
scikit-learn/scikit-learn: doc/developers/callback_support.rst

.. _callbacks_support:

===========================
Supporting the Callback API
===========================

.. currentmodule:: sklearn.callback

Scikit-learn offers a callback API to use built-in or custom callbacks with compatible
estimators. This section is intended for developers who wish to implement callback
support in estimators.

A callback is an object that can be registered to an estimator. The callback holds
methods, called hooks, which will get called at different specific steps during the
fitting process of the estimator it is attached to.

There are two types of callbacks : propagated and non-propagated. When using
meta-estimators, a propagated callback must be registered to the outermost
meta-estimator, and it will be propagated to the sub-estimators, meaning its hooks will
be called in both the :term:`fit` function of the meta-estimator and the ones of its
sub-estimators. By contrast a non-propagated callback can be registered to any meta- or

> sub-estimator, and its hooks will only be called in that estimator's ``fit`` function.

Reviewer: These 2 paragraph should more or less be in the user guide, they target users mostly. Instead I'd add a link to the user guide here. Maybe add a TODO until we have this user guide.

Supporting the callback API in an estimator boils down to enabling the registration of
callbacks, defining the different steps (referred to as "tasks") of its ``fit``
method, and calling the right hooks at the right time for each defined task. Doing so

> requires the use of three types of objects:

Reviewer: I'd rather not mention "hook" here since it hasn't been defined yet.
```suggestion
Adding callback support in an estimator boild down to enabling the registration of
callbacks, modelling :term:`fit` as a tree of tasks, and invoking the callbacks at the
beginning and end of each of these tasks. To integrate this, scikit-learn provides the
following helpers from the :mod:`~sklearn.callback` module:
```

Author: Two paragraph above it is mentioned that the callback methods are called hooks, but if that paragraph is moved to the user guide then this mention won't be there anymore. Then I wonder if we can assume that someone reading this page has read the user guide and directly use all the vocabulary defined there.

Reviewer: I wouldn't mention 'hooks' neither in this page nor in the user guide. Only in the developing callbacks page.

- :class:`~CallbackContext` objects which are responsible for calling the callback hooks
  and also holding contextual information necessary for the callbacks.

- the :class:`~CallbackSupportMixin` class from which the estimator must inherit in
  order to be able to register callbacks.

- the :func:`~with_callbacks` decorator, to decorate the estimator's ``fit`` method to

>   take care of the callbacks teardown.

Reviewer: I think that we should put the mixin first since it's the first thing to do
```suggestion
- :class:`~CallbackSupportMixin`, which enables callback registration and initializes
  callback handling at the beginning of fit.

- :class:`~CallbackContext`, which represents tasks and is the central object to manage
  callbacks during fit.

- :func:`~with_callbacks`, to guarantee proper callback tear down at the end of fit.
```

CallbackContext
---------------

The :class:`~CallbackContext` objects are responsible for calling the callback hooks at
the right time. They track the different tasks of the estimator, with one context object

> per task, and they capture the tree structure of the tasks of the estimator.

Reviewer: ```suggestion
The :class:`~CallbackContext` objects are responsible for invoking the callbacks at the
right time during fit. They track the different tasks of the estimator, with one context
representing each task, and capture the tree structure of the tasks involved in the
execution of the fit method.
```

> The task tree

Reviewer: Before introducing the task tree I would first explain what a task is. That is any unit of work defined by the estimator. In scikit-learn usually iterations of the learning algorithm, or steps of a pipeline, cross validation folds, ...

> ^^^^^^^^^^^^^

Reviewer: Instead of a sub-section I think that the description of the task tree would be better in a dropdown. above that I'd see a small paragraph to quickly explain how fit is decomposed into nested tasks, something like "Tasks (and therefore callback contexts) have a natural tree structure, where each task can be decomposed into subtasks and so on, with the root task being the whole fit task" (taken from the slep)

sub-estimator. Therefore, this leaf and the root task of the sub-estimator actually
represent the same task. In this case the leaf task of the meta-estimator and the root
task of the sub-estimator are merged into a single task.

For instance, a `Pipeline` would have a task tree that looks like this::

    Pipeline fit (root)
    ├── step 0 | StandardScaler fit
    │   └── <insert StandardScaler task tree here>
    └── step 1 | LogisticRegression fit
        └── <insert LogisticRegression task tree here>

In that case, a callback attached to the meta-estimator will have its hooks called
either only in the task nodes of the met-estimator (here ``fit``, ``setp 1`` and ``step
2``) if the callback is not propagated; or in both the tasks of the meta-estimator and
those of its sub-estimators (here the tasks of the ``StandardScaler`` and
``LogisticRegression`` task trees) if the callback is propagated.

The callback context's methods
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The :class:`~CallbackContext` objects are to be used through four methods :

.. method:: CallbackContext.call_on_fit_task_begin(estimator, **kwargs) -> None

>     :noindex:

Reviewer: I find that these method directives don't render well. Maybe use python code blocks instead.

In addition I think that we should put all the valid kwargs in the signature. I know that it's implemented as **kwargs but since we have a list of valid kwargs it's as if all kwargs were explicit.
We should also describe them here instead of in a later subsection. I would group the 2 call_on_fit_task_* methods in the same block to group the explanations.

(by the way, I think that we should make the signature explicit in the implem. Since we already do ``kwargs.get(param, None)``, we can default them all to None, it will be the same behavior.)

    arguments to provide extra information to be used by the callbacks. The list of
    possible keys and values for these kwargs is described in the :ref:`related section
    <callback_context_kwargs>` below.

.. method:: CallbackContext.call_on_fit_task_end(estimator, **kwargs) -> bool
    :noindex:

    This method must be called at the end of the context's task and wil call the
    callbacks' :meth:`~FitCallback.on_fit_task_end` hooks. Each of these hooks returns a
    boolean to request the ``fit`` process to stop, and this method will return ``True``
    if any of the hooks does. The estimator may thus use this boolean value to condition
    the interruption of its fitting.

    The arguments are the estimator of the current task, and the same options for
    additional keyword arguments as :meth:`~FitCallback.on_fit_task_begin`, listed in
    the :ref:`related section <callback_context_kwargs>` below.

.. method:: CallbackContext.subcontext(task_name="", task_id=None, max_subtasks=0, sequential_subtasks=True) -> CallbackContext instance
    :noindex:

    This method is used to generate a sub-context for a sub-task of the task the current

>     context is responsible for.

Reviewer: I find this sentence a bit circonvoluted :) maybe we can find a better formulation

    A string can be given to name the sub-task through the ``task_name`` argument, a
    unique id can be passed as the ``task_id`` argument.

    If the sub-task will also have sub-tasks, and if the maximum number of children
    tasks this sub-task can have is known, it must be indicated as an int with the
    ``max_subtasks`` argument. If it cannot be known in advance, ``max_subtasks`` must
    be set to ``None``. The ``sequential_subtasks`` argument is a boolean indicating if
    the sub-task's children tasks are sequential, and if so, the id of the children will
    be automatically generated, thus the ``task_id`` argument will need to be left to

>     ``None`` for these children tasks.

Reviewer: Here as well I would not detail all the parameters, they are already described in the docstring.
```suggestion
```

>     The generated sub-context is returned.

Reviewer: ```suggestion
    The generated sub-context is returned, which can then be used to create sub-sub-contexts, and so on.
```

.. method:: CallbackContext.propagate_callback_context(sub_estimator) -> self
    :noindex:

    This method is used in meta-estimators to propagate the callbacks to a sub-estimator.
    It will also propagate the context, merging it with the root context of the
    sub-estimator's ``fit`` function, as the current context's task and the

>     sub-estimator's root task represent the same task.

Reviewer: I'd first say that this method should be used on a context corresponding to the task of fitting a sub-estimator, making it a leaf task of the meta-estimator.

>     sub-estimator's root task represent the same task.

Reviewer: I'd add that therefore when composing estimators, the trees of individual estimators are combined into a single tree, rooted at the fit of the top level estimator.

    The current context is returned.

CallbackSupportMixin

> --------------------

Reviewer: As said above, let's move this first.

To support callbacks, an estimator must inherit from the :class:`~CallbackSupportMixin`
mixin, which provides the :meth:`~CallbackSupportMixin.set_callbacks` method to register

> callbacks to the estimator.

Reviewer: ```suggestion
To support callbacks, an estimator must inherit from the :class:`~CallbackSupportMixin`
class, which exposes the following methods:

- :meth:`~CallbackSupportMixin.set_callbacks`, a public method to be called by the user
  to register callbacks on the estimator.
```

It also provides the following method to initialize the root callback
context in ``fit`` :

.. method:: CallbackSupportMixin._init_callback_context(task_name="fit", task_id=0, max_subtasks=0, sequential_subtasks=True) -> CallbackContext instance
    :noindex:

    This method must be used at the beginning of the estimator's ``fit`` to generate the

>     callback context corresponding to the root task of the estimator's ``fit``.

Reviewer: ```suggestion
- :meth:`~CallbackSupportMixin._init_callback_context`, which should be called at the
  beginning of fit to create the root `~CallbackContext`. This context represents the task
  of running fit as a whole.
```

Reviewer: I would add a small note here (maybe a note directive will make too highlighted) to explain that this method is public but with a leading underscore to not appear in the auto-completion of the methods of the estimator since it's for internal usage, not for the user.

Author: While adding the note I realised that since it has the leading underscore, the method doctring is not rendered on the doc page for the `CallbackSupportMixin` page and thus the method name cannot be clicked on to check the signature adnd docstring. So I guess the signature and docstring should be mentioned here, or maybe there is a way to tell sphinx to show the method in the doc anyway.

Reviewer: something told me that you can try to add the following to the docstring of CallbackSupportMixin:
``.. automethod:: sklearn.callback.CallbackSupportMixin._init_callback_context``

Author: I found out how to make sphinx show it anyway.

    A string can be given to name the root task through the ``task_name`` argument and a
    unique id can be passed as the ``task_id`` argument. The maximum number of children
    tasks this root task can have must be indicated as an int with the ``max_subtasks``
    argument if it is known. If unknown, ``max_subtasks`` must be set to ``None``. The
    ``sequential_subtasks`` argument is a boolean indicating if this root task's
    children tasks are sequential, and if so, the id of the children will be
    automatically generated. Thus the ``task_id`` argument in the ``subcontext`` method

>     used to generate the contexts for these tasks will need to be left to ``None``.

Reviewer: I don't think that we need to detail the parameters of this method here. We already link to the docstring which describes each parameter.
```suggestion
```

    The generated :class:`~CallbackContext` instance is returned.

The ``with_callbacks`` decorator
--------------------------------

The ``fit`` function of the estimator must be decorated with the :func:`~with_callbacks`
decorator. This decorator takes care of calling the
:meth:`~_base._BaseCallback.teardown` hook of the callbacks after ``fit`` finishes, and
does so in a ``try`` / ``finally`` block, in order to always perform the callbacks

> teardown, even if ``fit`` exits on an error.

Reviewer: instead of mentionning the hook, I'd just say that it takes care of tearing down the callbacks.

Please also add that for scikit-learn estimators, the ``_fit_context`` decorator already does that and so ``with_callbacks`` should not be used.

Minimal example
---------------

Here is a minimal example of a custom estimator supporting callbacks::

    from sklearn.base import BaseEstimator
    from sklearn.callback import CallbackSupportMixin, with_callbacks

    class MyEstimator(CallbackSupportMixin, BaseEstimator):
        def __init__(self, max_iter):
            self.max_iter = max_iter

        @with_callbacks
        def fit(self, X, y):
            callback_ctx = self._init_callback_context(max_subtasks=self.max_iter)
            callback_ctx.call_on_fit_task_begin(estimator=self, X=X, y=y)

            for i in range(self.max_iter):
                subcontext = callback_ctx.subcontext(task_name="iteration")
                subcontext.call_on_fit_task_begin(estimator=self, X=X, y=y)

                # Do something

                subcontext.call_on_fit_task_end(estimator=self, X=X, y=y)

            callback_ctx.call_on_fit_task_end(estimator=self, X=X, y=y)

            return self

For a more detailed example of how to make a custom estimator or meta-estimator
compatible with scikit-learn's callback API, you can refer to this example :
:ref:`sphx_glr_auto_examples_callbacks_plot_callback_support.py`.

> 

Reviewer: To ease the review a bit, please extract this example into a separate PR and put a TODO here instead

.. _callback_context_kwargs:

Contextual information passed to callbacks
------------------------------------------

Additional contextual information can be forwarded to the callback hooks from the
:meth:`~CallbackContext.call_on_fit_task_begin` and
:meth:`~CallbackContext.call_on_fit_task_end` methods, taking the form optional keyword
arguments given to these methods. The content of these ``kwargs`` will depend on the
estimator's implementation, as an estimator might not be able to produce all the
possible values. The list of the possible keys and corresponding values for these
``kwargs`` is as follows.

- `X`: array-like
    The training data of the estimator.

- `y`: array-like
    The training targets of the estimator.

- `metadata`: dict

>     A dictionary containing the training and validation metadata.

Reviewer: let's forget validation for now.

- `reconstruction_attributes`: dict
    A dictionary of the fitted attributes of the estimator, used by the callback context
    to generate a `fitted_estimator`, i.e. an estimator instance ready to predict,
    transform, etc ... as if the fit had stopped at the end of this task. The
    `fitted_estimator` is the object which will be passed to the callback hooks if
    required.

For each of these kwargs, a callable can be provided instead of the actual value. When
it is the case, if a callback hook requires the kwarg, the callback context will execute
the callable and forward the returned value to the hook. This mechanism enables
lazy-loading the kwarg values, to avoid potentially costly computations when no callback

> require a kwarg value.

Reviewer: I'd see this paragraph as a dropdown

Here is a list of all the built-in callbacks in scikit-learn, and which ones of these
``kwargs`` must be given to the :class:`~CallbackContext` methods for the callback
to function.

============================== ====================== =========================== ====================
Callback                       Description            Required kwargs for         Required kwargs for
                                                      call_on_fit_task_begin      call_on_fit_task_end
============================== ====================== =========================== ====================
:class:`~ProgressBar`          Display progress bars. None                        None

> ============================== ====================== =========================== ====================

Reviewer: I don't think that this table is very useful here. I think its place should be in the user guide (without the 3rd column)

Author: I thought that if I were a dev updating my custom estimator to support callback, and if providing one of the kwarg was doable but at the cost of significant work, it would help deciding whether to bother doing it or not by checking which callbacks require that kwarg.

Reviewer: It's misleading: one might think that the required kwargs should be for every call to these functions, while in fact it's only when the estimator can provide that. That's what my point was.

Maybe instead of saying "required kwargs" we find a way to say kwargs that the callback can use.

Author: Since the kwargs description moved to the callback methods section, I no longer have a good place to put this table, I don't think it is worth having its own section, so I'm ok removing it here and just having it in the user guide.
